runtime/connectors/blob/objectreader.go

```diff
@@ -0,0 +1,94 @@
+package blob
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync/atomic"
+
+	"github.com/c2h5oh/datasize"
+	"gocloud.dev/blob"
+)
+
+// ObjectReader reads range of bytes from cloud objects
+// implements io.ReaderAt and io.Seeker interfaces
+type ObjectReader struct {
+	ctx    context.Context
+	bucket *blob.Bucket
+	index  int64
+	obj    *blob.ListObject
+
+	// debug data
+	debugMode bool
+	call      int64
+	bytes     int64
+}
+
+func (f *ObjectReader) ReadAt(p []byte, off int64) (int, error) {
+	if f.debugMode {
+		fmt.Printf("reading %v bytes at offset %v\n", len(p), off)
+		atomic.AddInt64(&f.call, 1)
+	}
+
+	reader, err := f.bucket.NewRangeReader(f.ctx, f.obj.Key, off, int64(len(p)), nil)
```

Comment:
I did think about it but didn't implement it for the sake of simplicity. There are quite a good number of edge cases particularly around handling of EOF like
1) client requesting more data then the buffer size in which case we need to bypass buffer
2) RangeReader returning less number of bytes than buffer size but more bytes than what is requested by client in which case NewRangeReader will return EOF but we should not return it to client. 
3) RangeReader returning less number of bytes than what is requested by client. In this case we should return EOF.
4) Reading from cache till the end of buffer in which case we need to return EOF based on what RangeReader returned. 

Plus we need to synchronise the access to the buffer since multiple columns are read in parallel. 

So considering all these cases I decided to not implement any caching. We can expect that in actual scenarios one column's data will mostly be higher than 4 MB (since recommended page size itself is 1MB and the reader reads entire column's data in one go).
In worst case we need to make number of calls equal to number of columns which is okay since calls are relatively cheap(from dollars point). The latency hit should be less since data is fetched in parallel.

Reply:
Okay, this sounds good for now if we're sure the Parquet/Arrow reading libraries will request large page sizes (several MB) when applicable and not small ones? I thought they were designed for on-disk data and would request maybe 4kb batches or something like that.

Reply:
We can also enable the option of reading data in batches rather than loading entire data in-memory. I didn't enable that since it was making many API calls.